src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolver.scala

```diff
@@ -0,0 +1,182 @@
+package io.iohk.ethereum.blockchain.sync.fast
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props, Scheduler}
+import io.iohk.ethereum.blockchain.sync.PeerRequestHandler.{RequestFailed, ResponseReceived}
+import io.iohk.ethereum.blockchain.sync.fast.FastSyncBranchResolver._
+import io.iohk.ethereum.blockchain.sync.{BlacklistSupport, PeerListSupport, PeerRequestHandler}
+import io.iohk.ethereum.db.storage.AppStateStorage
+import io.iohk.ethereum.domain.{BlockHeader, Blockchain}
+import io.iohk.ethereum.network.EtcPeerManagerActor.PeerInfo
+import io.iohk.ethereum.network.Peer
+import io.iohk.ethereum.network.p2p.messages.Codes
+import io.iohk.ethereum.network.p2p.messages.PV62.{BlockHeaders, GetBlockHeaders}
+import io.iohk.ethereum.utils.Config.SyncConfig
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+
+class FastSyncBranchResolver(
+    val fastSync: ActorRef,
+    val peerEventBus: ActorRef,
+    val etcPeerManager: ActorRef,
+    val blockchain: Blockchain,
+    val syncConfig: SyncConfig,
+    val appStateStorage: AppStateStorage,
+    implicit val scheduler: Scheduler
+) extends Actor
+    with ActorLogging
+    with PeerListSupport
+    with BlacklistSupport {
+
+  import syncConfig._
+
+  def handleCommonMessages: Receive = handlePeerListMessages orElse handleBlacklistMessages
+
+  override def receive: Receive = handleCommonMessages orElse { case StartBranchResolver =>
+    getBestPeer match {
+      case Some((peer, _)) => requestBlockHeadersGap(peer, blockchain.getBestBlockNumber())
+      case None =>
+        log.info(s"Still waiting for some peer, rescheduling StartBranchResolver")
+        scheduler.scheduleOnce(1.second, self, StartBranchResolver)
+    }
+  }
+
+  private def waitingBlockHeadersGap(bestBlockNumber: BigInt): Receive = handleCommonMessages orElse {
+    case ResponseReceived(peer, BlockHeaders(blockHeaders), timeTaken) => {
+      log.info("*** Received {} block headers in {} ms ***", blockHeaders.size, timeTaken)
+      getLastCorrectBlock(blockHeaders, bestBlockNumber) match {
+        case Some(lastValidBlockHeader) =>
+          finalizeBranchResolver(bestBlockNumber, lastValidBlockHeader)
+        case None => requestBlockHeader(peer, SearchState(0, bestBlockNumber))
+      }
+    }
+    case RequestFailed(peer, reason) => handleRequestFailure(peer, sender(), reason)
+  }
+
+  private def binarySearchingLastValidBlock(searchState: SearchState): Receive = handleCommonMessages orElse {
+    case ResponseReceived(peer, BlockHeaders(blockHeaders), timeTaken) =>
```

Comment:
during binary seach we only ask for one header from potential master peer, so we should validate if:
 - response has exactly one header
 - response is from master peer